File: algorithms/2_medium/K/00875_koko_eating_bananas/binary_search.c
int minEatingSpeed(const int *restrict piles, const int len, const int h) {
    if (piles == NULL || len <= 0 || h <= 0 || len > h) { return -1; }

    int hi = 0;
    for (int i = 0; i < len; i += 1) { hi = MAX(hi, piles[i]); }
    int lo = 1;

    while (lo <= hi) {
        const int mid = lo + (hi - lo) / 2;

        int total = 0;
        for (int i = 0; i < len; i += 1) {
            total += (piles[i] + mid - 1) / mid;
            if (total > h) {
                break;
            }
        }

        if (total <= h) {
            hi = mid - 1;
        } else {
            lo = mid + 1;
        }
    }

    return lo;
}

File: algorithms/2_medium/K/00875_koko_eating_bananas/binary_search.rs
impl Solution {
    pub fn min_eating_speed(piles: Vec<i32>, h: i32) -> i32 {
        let mut hi: i32 = *piles.iter().max().unwrap();
        let mut lo: i32 = 1;

        while lo <= hi {
            let mid = lo + (hi - lo) / 2;

            let mut total = 0_i32;
            for &p in &piles {
                total += (p + mid - 1) / mid;
                if total > h {
                    break;
                }
            }

            if total <= h {
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        }

        lo
    }
}
